### PR TITLE
MG-391: Add y_axis_max to immunohisto data

### PR DESCRIPTION
**Pull request by beatrizsaldana**
## Problem
The immunohisto transform was missing y-axis maximum values needed for proper visualization of data. Without these values, frontend applications couldn't properly scale figures, leading to poor user experience when viewing biomarker and pathology data.

## Solution
This PR implements a comprehensive solution for calculating and rounding y-axis maximum values in the immunohisto transform:

**Key Features Added:**
1. **Y-axis Maximum Calculation**: Added logic to calculate the maximum value across all ages for each combination of `(name, evidence_type, tissue)`
2. **Smart Rounding Algorithm**: Implemented `round_y_axis_max()` function with sophisticated rounding logic that:
   - Returns 10 for zero values
   - Rounds UP to the next "nice" number where the second digit is 0 or 5
   - Handles different magnitudes correctly (e.g., 0.0021 → 0.0025, 1094 → 1500)
   - Maintains monotonicity and handles floating-point precision issues
3. **Comprehensive Error Handling**: Handles edge cases like negative values, conversion errors, and missing data

**Rounding Logic Details:**
The rounding algorithm follows a specific pattern with key design principles:
1. Always Round UP: Ensures the y-axis scale accommodates all data points
2. "Nice Numbers": Creates visually appealing scales with 0 or 5 as second digit
3. Magnitude Preservation: Maintains the order of magnitude while rounding
4. Floating-Point Safety: Uses string manipulation to avoid precision issues
5. Monotonicity: Larger inputs always produce larger or equal outputs

## Tests
Added comprehensive test coverage including:
- **Unit tests** for `round_y_axis_max()` function with 15+ test cases covering:
  - All Jira ticket examples
  - Edge cases (zero, negative values, very small/large numbers)
  - Second digit rounding logic
  - Magnitude handling
  - Floating-point precision issues
  - Monotonicity verification
  - Parametrized tests for main examples
- **Integration tests** updated to include expected `y_axis_max` field in all test output files
- **Test assets** updated with new expected output containing y-axis max values

**Files Modified:**
- `src/agoradatatools/etl/transform/immunohisto_transform.py` - Core implementation
- `tests/transform/test_immunohisto_transform.py` - Comprehensive test suite
- All test output JSON files updated with expected y_axis_max values

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -35,6 +35,7 @@ install_requires =
     PyYAML~=6.0
     pyarrow~=14.0.1
     typer~=0.7.0
+    click<8.3.0
```
Comment (beatrizsaldana) [inline setup.cfg]: To prevent CI failure

File: tests/transform/test_immunohisto_transform.py [hunk 2]
```diff
@@ -319,3 +474,537 @@ def test_extract_age_num_mixed_error_handling(self):
...
+        assert "12 months" in ages_found
+
+        # Check common properties
+        for entry in result:
+            assert entry["name"] == "Model1"
+            assert entry["evidence_type"] == "Type1"
+            assert entry["tissue"] == "Tissue1"
+            assert entry["units"] == "mg"
+            # round_y_axis_max(3.0) returns 3.5 (rounds up to next nice number)
+            assert entry["y_axis_max"] == pytest.approx(3.5)
```
[the remark this answers is not in the document]
Comment (beatrizsaldana) [reply tests/transform/test_immunohisto_transform.py]: This is a test and we want to assert that the function provided the expected result. I don't think this suggestion is useful.

File: src/agoradatatools/etl/transform/immunohisto_transform.py [hunk 3]
```diff
@@ -65,6 +66,231 @@ def prepare_immunohisto_data(df: pd.DataFrame) -> pd.DataFrame:
...
+
+    # Handle special cases: zero or negative values
+    if y_axis_max <= 0:
+        return 10.0 if y_axis_max == 0 else 0.0
+
+    # Find the order of magnitude of the number
+    magnitude = int(math.floor(math.log10(y_axis_max)))
+
+    # Scale the number so the first digit is in the ones place
+    scaled = y_axis_max / (10**magnitude)
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/immunohisto_transform.py]: I think you can bypass all of the code below by doing this instead:
```
scaled = y_axis_max / (10**(magnitude-1)) # 1st and 2nd digits are now in the tens and 1s place
rounded_scaled = 5 * math.ceil(scaled / 5) # Round up to the nearest '5'. 
# If the second digit of scaled is '0', rounded_scaled will stay the same as scaled and needs to get 5 added
if rounded_scaled == int(scaled):
  rounded_scaled = rounded_scaled + 5

# Put back to the right magnitude
result = rounded_scaled * (10**(magnitude - 1))

# remove float precision issues
result = round(result, 15)
```

I tested this with every case listed in the comments above and it seems to work. I also tested it with 10 (magnitude 1) to make sure 1-1 didn't cause issues. This will _not_ work for numbers <= 10^-15 but I'm assuming we wouldn't have a value that small. We could probably add a check for if y_axis_max <= 10^-10, return 10, or something. Or you can keep your precision checking code and round based on max(15, precision).

File: src/agoradatatools/etl/transform/immunohisto_transform.py [hunk 4]
```diff
@@ -65,6 +66,231 @@ def prepare_immunohisto_data(df: pd.DataFrame) -> pd.DataFrame:
...
+
+
+def _calculate_y_axis_max_map(
+    dataset: pd.DataFrame,
+) -> Dict[Tuple[str, str, str], float]:
+    """
+    Calculate y_axis_max for each combination of (name, evidence_type, tissue) across all ages.
+
+    Args:
+        dataset: The prepared dataset
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/immunohisto_transform.py]:
```suggestion
        dataset: The prepared dataset. It must have columns "name", "evidence_type", "tissue", and "value".
```
Since this data frame has already been validated to have these columns in the main transform function, I don't think we need to add error checking here, but I think it's helpful if the comments explicitly say what columns are used here.

File: src/agoradatatools/etl/transform/immunohisto_transform.py [hunk 5]
```diff
@@ -65,6 +66,231 @@ def prepare_immunohisto_data(df: pd.DataFrame) -> pd.DataFrame:
...
+    result = (first_digit + rounded_second / 10.0) * (10**magnitude)
+
+    return result
+
+
+def _calculate_y_axis_max_map(
+    dataset: pd.DataFrame,
+) -> Dict[Tuple[str, str, str], float]:
+    """
+    Calculate y_axis_max for each combination of (name, evidence_type, tissue) across all ages.
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/immunohisto_transform.py]: Nit: Per my comment below, this function is confusingly named and documented if it's not actually calculating the final y_axis_max value with `round_y_axis_max`.

File: src/agoradatatools/etl/transform/immunohisto_transform.py [hunk 6]
```diff
@@ -65,6 +66,231 @@ def prepare_immunohisto_data(df: pd.DataFrame) -> pd.DataFrame:
...
+    data_rows = []
+    grouped = dataset.groupby(group_columns)
+
+    for group_key, group in grouped:
+        entry = dict(zip(group_columns, group_key))
+
+        # Get the y_axis_max for this combination of (name, evidence_type, tissue)
+        key_for_y_axis = (entry["name"], entry["evidence_type"], entry["tissue"])
+        raw_y_axis_max = y_axis_max_map.get(key_for_y_axis, 0)
+        entry["y_axis_max"] = round_y_axis_max(raw_y_axis_max)
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/immunohisto_transform.py]: Nit: Is there a reason you don't call `round_y_axis_max` in the `_calculate_y_axis_max_map` function instead and treat the lookup table as a table of final values? Otherwise it doesn't make sense why the lookup table needs a separate function if it doesn't actually have the final values you want.

File: src/agoradatatools/etl/transform/immunohisto_transform.py [hunk 7]
```diff
@@ -65,6 +66,231 @@ def prepare_immunohisto_data(df: pd.DataFrame) -> pd.DataFrame:
...
+
+    Args:
+        dataset: The prepared dataset
+        group_columns: Columns to group by
+        extra_columns: Columns to include in the data
+        extra_column_name: Name of the extra column
+        y_axis_max_map: Pre-calculated y_axis_max values
+
+    Returns:
+        List of data rows with y_axis_max values
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/immunohisto_transform.py]: Nit: This docstring needs a few more details. What is meant by "data rows"? `extra_column_name` needs a more specific description too. `extra_columns` should say something like "other columns in addition to group_columns to include in the returned data"

File: src/agoradatatools/etl/transform/immunohisto_transform.py [hunk 8]
```diff
@@ -65,6 +66,231 @@ def prepare_immunohisto_data(df: pd.DataFrame) -> pd.DataFrame:
...
+        dataset: The prepared dataset
+
+    Returns:
+        Dictionary mapping (name, evidence_type, tissue) tuples to their max values
+    """
+    key_dimensions = ["name", "evidence_type", "tissue"]
+    y_axis_max_map = {}
+
+    for key, group in dataset.groupby(key_dimensions):
+        if len(group) > 0:
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/immunohisto_transform.py]: I'm curious, is `len(group)` ever 0?

File: src/agoradatatools/etl/transform/immunohisto_transform.py [hunk 9]
```diff
@@ -65,6 +66,231 @@ def prepare_immunohisto_data(df: pd.DataFrame) -> pd.DataFrame:
...
+        data_rows.append(entry)
+
+    return data_rows
+
+
+def _add_missing_age_entries(
+    data_rows: List[Dict[str, Any]], dataset: pd.DataFrame
+) -> List[Dict[str, Any]]:
+    """
+    Add placeholder entries for missing age combinations to ensure data completeness.
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/immunohisto_transform.py:152]: Can you add an example to the comments to show what this function is doing? Something like, if the data set has ages 4, 8, and 12 months but one tissue is missing 8 month data, this function adds <XYZ> to data_rows to fill in the missing time point.

File: src/agoradatatools/etl/transform/immunohisto_transform.py [hunk 10]
```diff
@@ -65,6 +66,231 @@ def prepare_immunohisto_data(df: pd.DataFrame) -> pd.DataFrame:
...
+        dataset: The original dataset
+
+    Returns:
+        Updated data_rows with missing age entries added
+    """
+    # Get all unique ages that exist in the dataset
+    available_ages = list(set([x["age"] for x in data_rows]))
+
+    # Create a lookup map for y_axis_max values from already processed entries
+    y_axis_max_lookup = {}
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/immunohisto_transform.py]: Related to comments above, if you create a lookup map with the final y axis values, you could just pass it in to this function instead of having to re-make it.

File: src/agoradatatools/etl/transform/immunohisto_transform.py [hunk 11]
```diff
@@ -65,6 +66,231 @@ def prepare_immunohisto_data(df: pd.DataFrame) -> pd.DataFrame:
...
+
+        # Get the y_axis_max for this combination of (name, evidence_type, tissue)
+        key_for_y_axis = (entry["name"], entry["evidence_type"], entry["tissue"])
+        raw_y_axis_max = y_axis_max_map.get(key_for_y_axis, 0)
+        entry["y_axis_max"] = round_y_axis_max(raw_y_axis_max)
+
+        entry[extra_column_name] = group[extra_columns].to_dict("records")
+        data_rows.append(entry)
+
+    return data_rows
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/immunohisto_transform.py]: A suggestion for simplifying this function. You might even be able to delete this a function and move this to the main code since it's short:
```
# Add nest_fields from agoradatatools.etl.utils to the imports at the top of the file

grouped = nest_fields(dataset.copy(), grouping=group_columns, new_column="data", drop_columns=group_columns)

grouped["y_axis_max"] = grouped.apply(
  lambda entry: 
    round_y_axis_max(y_axis_max_map.get((entry["name"], entry["evidence_type"], entry["tissue"]), 0)), axis=1)
  # Though as above, I really think the rounding should already be in the lookup table
))

return grouped.to_dict("records")
```

File: src/agoradatatools/etl/transform/immunohisto_transform.py [hunk 12]
```diff
@@ -65,6 +66,231 @@ def prepare_immunohisto_data(df: pd.DataFrame) -> pd.DataFrame:
...
+                        "evidence_type": entry["evidence_type"],
+                        "tissue": entry["tissue"],
+                        "age": age,
+                        "units": "",  # Empty units since there's no actual data
+                        "y_axis_max": y_axis_max,
+                        "data": [],  # Empty data array since there are no measurements for this age
+                    }
+                )
+
+    return data_rows
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/immunohisto_transform.py]: A suggestion for simplifying this function. If you use my suggestion for _create_data_rows but leave data_rows as a data frame, you can play some games with pandas to fill in missing ages. This removes the need to pass in the original data set or create a y_max lookup table and you can just return the data frame and I don't think you need to call convert_numpy_types either. 
```
# add import numpy as np to imports at top of function

available_ages = list(data_rows["age"].drop_duplicates())
# All unique combinations of groups
fill_df = data_rows[["name", "evidence_type", "tissue", "y_axis_max"]].copy().drop_duplicates()
fill_df["age"] = [available_ages] * fill_df.shape[0] # Make an "age" column where each entry is a list of all possible ages

# "explode" makes one row per age + group. Then merge back into the data to create new rows for missing ages
fill_df = fill_df.explode("age").merge(data_rows, how="outer", validate="one_to_one") 

# Fill NA values for units. Can't use fillna to make an empty list so we add an extra line
fill_df = fill_df.fillna({"units": ""})
fill_df["data"] = fill_df["data"].apply(lambda x: [] if x is np.NaN else x)

# sort by age
# Adjust _extract_age_num to work with a single string from a pd.Series, no need to return a tuple
def _extract_age_num(entry: str) -> float:
  try/catch but return float

fill_df["age_numeric"] = fill_df["age"].apply(_extract_age_num)

# Sort
fill_df = fill_df.sort_values(["age_numeric", "age", "evidence_type"]).drop(columns="age_numeric") 
```

File: tests/transform/test_immunohisto_transform.py [hunk 13]
```diff
@@ -4,8 +4,13 @@
 import pytest
```
Comment (jaclynbeck-sage) [inline tests/transform/test_immunohisto_transform.py:16]: I'll review the tests after you simplify. You might be able to remove some of the helper functions and delete/simplify those tests.

**Review by jaclynbeck-sage — CHANGES_REQUESTED** [review]
This is a clever solution to making sure things get rounded correctly. I do think this code can be simplified a lot and I made some suggestions on how to do that with pandas instead of lists of dictionaries.

File: src/agoradatatools/etl/transform/immunohisto_transform.py (hunk 3 above)
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/immunohisto_transform.py]: I implemented this with a minor change, thank you!

File: src/agoradatatools/etl/transform/immunohisto_transform.py (hunk 8 above)
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/immunohisto_transform.py]: The check is defensive but unnecessary. I looked into it and I think pandas groupby only yields groups that contain data, soooo len(group) will always be greater than 0. For some reason I am afraid of removing it. But I will. We do not need redundant stuff in this already very complicated transform.

File: src/agoradatatools/etl/transform/immunohisto_transform.py [hunk 14]
```diff
@@ -65,6 +67,149 @@ def prepare_immunohisto_data(df: pd.DataFrame) -> pd.DataFrame:
...
+
+    Returns:
+        float: The numeric age value, or inf for invalid ages
+    """
+    try:
+        numeric_value = float(age_str.split()[0])
+        return numeric_value
+    except (ValueError, IndexError, AttributeError):
+        # For invalid ages, use a large number for consistent sorting
+        return float("inf")
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/immunohisto_transform.py]: To double-check: @JessterB if the age field is mis-formed (i.e. "6months" instead of "6 months") or missing values, do we want to throw an error, or return `inf` as written? I'm a little worried about it failing silently with `inf` and us not catching it.

Comment (JessterB) [reply src/agoradatatools/etl/transform/immunohisto_transform.py]: Throwing an error here seems reasonable; I'd also expect GX validation to fail on malformed age strings, so I don't expect we would miss the issue entirely once we add GX suites.

Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/immunohisto_transform.py]: I'll throw an error! I did the inf so that we could still get a result and use age for sorting.

**Review by jaclynbeck-sage — APPROVED** [review]
Looks great! As usual, amazing job with the thorough testing and error checking.

**Review by JessterB — APPROVED** [review]
Ship it!
